compiler/planner/insert_planner.rs

```diff
@@ -0,0 +1,402 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+// jk there's no planning to be done here, just execution.
+// There is a need to construct the executor though.
+
+use std::{
+    collections::{BTreeSet, HashMap, HashSet},
+    error::Error,
+    fmt::{Display, Formatter},
+    marker::PhantomData,
+    sync::Arc,
+};
+
+use answer::{variable::Variable, variable_value::VariableValue, Type};
+use concept::type_::type_manager::TypeManager;
+use encoding::{
+    graph::type_::{CapabilityKind::Plays, Kind},
+    value::{label::Label, value::Value},
+};
+use ir::pattern::{
+    constraint::{Comparison, Constraint, Constraints, Isa, RoleName},
+    expression::Expression,
+};
+use storage::snapshot::{ReadableSnapshot, WritableSnapshot};
+
+use crate::inference::type_annotations::TypeAnnotations;
+
+macro_rules! filter_variants {
+    ($variant:path : $iterable:expr) => {
+        $iterable.iter().filter_map(|item| if let $variant(inner) = item { Some(inner) } else { None })
+    };
+}
+
+pub type VariablePosition = usize; // Why is that not in plan.
+
+#[derive(Debug, Clone, Hash, Ord, PartialOrd, Eq, PartialEq)]
+pub enum VariableSource {
+    TypeSource(TypeSource),
+    ValueSource(ValueSource),
+    ThingSource(ThingSource),
+}
```

Comment:
Similar purpose as `VariablePosition`, but wrapped to indicate which array to look in.